src/app/icd/client/ICDRefreshKeyInfo.h

```diff
@@ -0,0 +1,134 @@
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandSender.h>
+#include <app/OperationalSessionSetup.h>
+
+#include <crypto/CHIPCryptoPAL.h>
+#include <lib/core/CHIPConfig.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/ScopedNodeId.h>
+#include <lib/support/CodeUtils.h>
+#include <stddef.h>
+
+namespace chip {
+namespace app {
+
+class CheckInDelegate;
+class ICDRefreshKeyInfo;
+
+/**
+ * @brief ICDRefreshKeyInfo contains all the data and methods needed for key refresh and re-registration of an ICD client.
+ */
+class ICDRefreshKeyInfo
+{
+public:
+    typedef Crypto::SensitiveDataBuffer<Crypto::kAES_CCM128_Key_Length> RefreshKeyBuffer;
+    /**
+     * @brief This class comprises command sender callbacks for the registration command.
+     */
+    class Callback : public CommandSender::Callback
+    {
+    public:
+        // CommandSender callbacks
+        void OnResponse(CommandSender * apCommandSender, const ConcreteCommandPath & aPath, const StatusIB & aStatus,
+                        chip::TLV::TLVReader * aData) override;
+        void OnError(const CommandSender * apCommandSender, CHIP_ERROR aError) override;
+        void OnDone(CommandSender * apCommandSender) override;
+
+        void SetICDRefreshKeyInfo(ICDRefreshKeyInfo * apICDRefreshKeyInfo) { mpICDRefreshKeyInfo = apICDRefreshKeyInfo; }
+
+        virtual ~Callback() {}
+
+    private:
+        CHIP_ERROR mError                       = CHIP_NO_ERROR;
+        ICDRefreshKeyInfo * mpICDRefreshKeyInfo = nullptr;
+    };
+
+    ICDRefreshKeyInfo(CheckInDelegate * apCheckInDelegate, ICDClientInfo & aICDClientInfo, ICDClientStorage * aICDClientStorage);
+
+    /**
+     * @brief Used by the application to set a new key to avoid counter rollover problems.
+     *
+     * @param[in] clientInfo clientInfo object
+     * @param[in] keyData New key data to use to re-register the client with the server
+     * @param[in] exchangeMgr exchange manager to use for the re-registration
+     * @param[in] sessionHandle session handle to use for the re-registration
+     */
+    CHIP_ERROR RegisterClientWithNewKey(Messaging::ExchangeManager & exchangeMgr, const SessionHandle & sessionHandle);
+
+    /**
+     * @brief Sets up a CASE session to the peer for re-registering a client with the peer when a key refresh is required to avoid
+     * ICD counter rollover. Returns error if we did not even manage to kick off a CASE attempt.
+     */
+    CHIP_ERROR EstablishSessionToPeer();
+
+    /**
+     * @brief Getter for ICDClientInfo
+     */
+    ICDClientInfo & GetICDClientInfo(void) { return mICDClientInfo; }
+
+    /**
+     * @brief Getter for ICDClientStorage pointer
+     */
+    ICDClientStorage * GetICDClientStorage(void) { return mpICDClientStorage; }
+
+    /**
+     * @brief Getter for CheckInDelegate pointer
+     */
+    CheckInDelegate * GetCheckInDelegate(void) { return mpCheckInDelegate; }
+
+    RefreshKeyBuffer mNewKey;
```

Comment:
public? should be private? Create a public function to init this new key?

Reply:
I had issues returning the SensitiveDataBuffer. Will try with a reference.

Reply:
Modified. Please re-review.